### ISSUE TITLE
It's not said about .appxrecipe for UWP test pipeline

### ISSUE THREAD

**Issue by aosyatnik** [MicrosoftDocs/vsts-docs#6108]
In the documentation it's mentioned only `**\*test*.dll`, but it's possible to use `.appxrecipe` for UWP test pipeline.
Although I still do not understand why it's failing in the build, but tests are running. It would be very helpful if you add an example of how to configure Visual Studio Test task for UWP.

Here I created a repo with an example: https://dev.azure.com/melashkina0755/UWP_App_With_Tests

Pipeline YAML:
```
# Universal Windows Platform
# Build a Universal Windows Platform project using Visual Studio.
# Add steps that test and distribute an app, save build artifacts, and more:
# https://aka.ms/yaml

trigger:
- master

pool:
  vmImage: 'windows-latest'

variables:
  solution: '**/*.sln'
  buildPlatform: 'x86|x64'
  buildConfiguration: 'Release'
  appxPackageDir: '$(build.artifactStagingDirectory)\AppxPackages\\'

steps:
- task: NuGetToolInstaller@1

- task: NuGetCommand@2
  inputs:
    restoreSolution: '$(solution)'

- task: VSBuild@1
  inputs:
    platform: 'x86'
    solution: '$(solution)'
    configuration: '$(buildConfiguration)'
    msbuildArgs: '/p:AppxBundlePlatforms="$(buildPlatform)" /p:AppxPackageDir="$(appxPackageDir)" /p:AppxBundle=Always /p:UapAppxPackageBuildMode=StoreUpload'

- task: VSTest@2
  inputs:
    platform: 'x86|x64'
    testSelector: 'testAssemblies'
    testAssemblyVer2: |
      **\UnitTestProject1.build.appxrecipe
      !**\*TestAdapter.dll
      !**\obj\**
    searchFolder: '$(System.DefaultWorkingDirectory)'
```

Tests are green: https://dev.azure.com/melashkina0755/UWP_App_With_Tests/_build/results?buildId=3&view=ms.vss-test-web.build-test-results-tab
Build itself failed: https://dev.azure.com/melashkina0755/UWP_App_With_Tests/_build/results?buildId=3

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 16bd8dd1-7650-8c18-2553-38a8bc54e892
* Version Independent ID: f9fa517a-7adc-e327-d12e-256295a1ed52
* Content: [Visual Studio Test task - Azure Pipelines](https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/test/vstest?view=azure-devops#q--a)
* Content Source: [docs/pipelines/tasks/test/vstest.md](https://github.com/MicrosoftDocs/vsts-docs/blob/master/docs/pipelines/tasks/test/vstest.md)
* Product: **devops**
* Technology: **devops-cicd**
* GitHub Login: @PBoraMSFT
* Microsoft Alias: **pbora**

**Comment by WilliamAntonRohm** [comment]
@aosyatnik -- Anna, thank you for your detailed feedback.  You may find some answers here:
- [Azure DevOps on Stack Overflow](https://stackoverflow.com/questions/tagged/azure-devops) 
- [Azure DevOps Support Bot](https://azuredevopsvirtualagent.azurewebsites.net/) 

 @PBoraMSFT -- Prachi, please look into this issue for potential documentation updates.

**Comment by PBoraMSFT** [comment]
@aosyatnik - I see in the pipeline sample that you have provided that you are using Microsoft hosted agents. Microsoft hosted agents are not equipped to run UWP workloads as they are Server OS images. 

There are multiple things to do here. 
1. Add an E2E UWP example in the documentation. Can I please request you to file a separate issue for that, so that all aspects of UWP development (and not just test) can be addressed in that?
2. Some tests seem to be succeeding even though the error messages seem to indicate that the app didn't start. That looks suspicious. Can you try running your pipeline on a self hosted agent? If you still run into issues with running the tests, pls loop back on this thread and we can take it forward.

**Comment by aosyatnik** [comment]
Hi, @PBoraMSFT , thank you for your reply.

2. I tried to run on self-hosted agent. I created another branch called `with_self_hosted_agent`
Here is the result: https://dev.azure.com/melashkina0755/UWP_App_With_Tests/_build/results?buildId=8

As self-hosted agent, I used my laptop.
![image](https://user-images.githubusercontent.com/8158404/67477628-cab72300-f65a-11e9-89a1-5d09019246dc.png)


Tests are not running at all (the same error that it couldn't connect to the test process), although build is successful.

**Comment by SadagopanRajaram** [comment]
@aosyatnik - I see that the yaml suggests that the build was done for x86 but the tests are run for x86 and x64. Can you please check to make sure that the architecture between the build and test task match?

**Comment by aosyatnik** [comment]
Hi @SadagopanRajaram , thank you for suggestions. I tried to run only x86. No good result.
Here is a build for Windows agent (1 test passed): https://dev.azure.com/melashkina0755/UWP_App_With_Tests/_build/results?buildId=9
Here is a build for Self-hosted agent (no test assed): https://dev.azure.com/melashkina0755/UWP_App_With_Tests/_build/results?buildId=10

I'm wondering why there is no good working example among all Microsoft examples? =\

**Comment by karanjitsingh** [comment]
@aosyatnik, from the logs it seems the task picks up two .appxrecipe files. Out of which the second one is failing. 

```
"d:\a\1\s\UnitTestProject1\bin\x86\Release\UnitTestProject1.build.appxrecipe"
"d:\a\1\s\UnitTestProject1\bin\x86\Release\Upload\UnitTestProject1.build.appxrecipe"
```

My guess is the second one comes from the parameter `/p:UapAppxPackageBuildMode=StoreUpload`. This could be happening because a `StoreUpload` version does not support deploying.

It seems a little unnecessary to run tests on both of these files. So you can either try to remove it or filter this particular appxrecipe file by giving this specific filter `**\Release\UnitTestProject1.build.appxrecipe`.

cc @mayankbansal0187

**Comment by aosyatnik** [comment]
Hi @karanjitsingh . Thank you it finally worked! I can see running test AND build didn't fail! Great work! 💃 
Here is a build: https://dev.azure.com/melashkina0755/UWP_App_With_Tests/_build/results?buildId=13&view=logs

@PBoraMSFT as build issue is resolved. We can think about adding it to documentation. I find it useful to have such an example for UWP. Where issue should be opened?

**Comment by aosyatnik** [comment]
For anyone who is having issues with the UWP test task. This is the right YML file:

```
# Universal Windows Platform
# Build a Universal Windows Platform project using Visual Studio.
# Add steps that test and distribute an app, save build artifacts, and more:
# https://aka.ms/yaml

trigger:
- master

pool:
  vmImage: 'windows-latest'

variables:
  solution: '**/*.sln'
  buildPlatform: 'x86|x64'
  buildConfiguration: 'Release'
  appxPackageDir: '$(build.artifactStagingDirectory)\AppxPackages\\'

steps:
- task: NuGetToolInstaller@1

- task: NuGetCommand@2
  inputs:
    restoreSolution: '$(solution)'

- task: VSBuild@1
  inputs:
    platform: 'x86'
    solution: '$(solution)'
    configuration: '$(buildConfiguration)'
    msbuildArgs: '/p:AppxBundlePlatforms="$(buildPlatform)" /p:AppxPackageDir="$(appxPackageDir)" /p:AppxBundle=Always /p:UapAppxPackageBuildMode=StoreUpload'

- task: VSTest@2
  inputs:
    platform: 'x86|x64'
    testSelector: 'testAssemblies'
    testAssemblyVer2: |
      **\Release\UnitTestProject1.build.appxrecipe <-- HERE IS FIX
      !**\*TestAdapter.dll
      !**\obj\**
    searchFolder: '$(System.DefaultWorkingDirectory)'
```

**Comment by PBoraMSFT** [comment]
@aosyatnik - thanks for the feedback and I am glad to see that Karan was able to help you get unblocked. I have filed 2 issues for additional documentation:

https://github.com/MicrosoftDocs/vsts-docs/issues/6240
https://github.com/MicrosoftDocs/vsts-docs/issues/6239

**Comment by mateialexandru** [comment]
@aosyatnik @PBoraMSFT thanks for this post ! Shed some light on uwp testing infrastructure.

Wanted to ask for clarification on the build agent: do we need a selfhosted agent ? Aka no way around hosted agent images just because they are windows server skus ?
If so, what extra steps did you have to take in configuring the selfhosted agent (assuming that vs, sdk is already installed ) : Is turning developer mode on enough ?

**Comment by aosyatnik** [comment]
Hi @mateialexandru .
I used ms agent, with a self-hosted agent it didn't work. SO you do not need a self-hosted agent. You might check my branch in repo here: https://dev.azure.com/melashkina0755/UWP_App_With_Tests/_git/81b0f53a-f8d9-46da-a794-36706ca7bbf5

There you will find `azure-pipelines.yml` file. There you can find
```
pool:
  vmImage: 'windows-latest'
```

This is ms agent:
![image](https://user-images.githubusercontent.com/8158404/70740065-a0a7e600-1d18-11ea-9d86-835dd344dabe.png)
